### ISSUE TITLE
Report unused packages

### ISSUE THREAD

**Issue by dnfield** [dart-lang/linter#2431]
Unused packages can be harmful in that they introduce additional network overhead and version solving constraint that isn't actually necessary.

For example, if I have a `dependency` on `package:foo`, but never `import 'package:foo/[*]';` anywhere in my code.

For bonus points, it would be nice if it could also catch that I have a `dependency` on `package:bar` but only ever use `import 'package:bar/bar.dart';` in test code.

**Comment by dnfield** [comment]
(sorry if this is a duplicate!)

**Comment by Hixie** [comment]
Some packages don't get referenced by code (e.g. asset packages) so we'd need a way for a package to opt-out of this probably.

**Comment by dnfield** [comment]
Yeah, another good reason to opt out would be if you're explicitly mentioning a transitive dependency for whatever reason, but never actually import it.
